scene/2d/visual_shape_2d.cpp

```diff
@@ -0,0 +1,351 @@
+	switch (shape_type) {
+		case SHAPE_RECTANGLE: {
+			points = {
+				Vector2(-1, 1),
+				Vector2(-1, -1),
+				Vector2(1, -1),
+				Vector2(1, 1)
+			};
+		} break;
+		case SHAPE_CIRCLE: {
+			float angle = Math_TAU / resolution;
+			points.resize(resolution);
+			for (int i = 0; i < resolution; i++) {
+				points.write[i] = Vector2(Math::sin(i * angle), -Math::cos(i * angle));
+			}
+		} break;
+		case SHAPE_EQUILATERAL_TRIANGLE: {
+			points = {
+				Vector2(-1, 1),
+				Vector2(0, -1),
+				Vector2(1, 1)
+			};
+		} break;
+		case SHAPE_RIGHT_TRIANGLE: {
+			points = {
+				Vector2(-1, 1),
+				Vector2(-1, -1),
+				Vector2(1, 1)
+			};
+		} break;
+		case SHAPE_CAPSULE: {
+			// Also see CapsuleShape2D::_get_points().
+			int capsule_res = resolution;
+			// Must be even.
+			if (resolution % 2 == 1) {
+				capsule_res += 1;
+			}
+			points.resize(capsule_res + 2);
+			int first_half = capsule_res / 2;
+			real_t turn_step = Math_TAU / capsule_res;
+			real_t radius = MIN(size.x, size.y);
+			real_t height = MAX(size.x, size.y);
+			if (radius == height) {
+				height += 0.0001;
+			}
+
+			Vector2 capsule_offset;
+			if (size.y >= size.x) {
+				capsule_offset = Vector2(0, 1.0 - radius / height);
+			} else {
+				capsule_offset = Vector2(1.0 - radius / height, 0);
+			}
+
+			int index = 0;
+			for (int i = 0; i < capsule_res; i++) {
+				Vector2 circle_point;
+				if (size.y >= size.x) {
+					// Start at right for vertical capsules.
+					circle_point = Vector2(Math::cos(i * turn_step), Math::sin(i * turn_step)) * radius / size;
+				} else {
+					// Start at bottom for horizontal capsules.
+					circle_point = Vector2(Math::sin(i * turn_step), Math::cos(i * turn_step)) * radius / size;
+				}
+
+				if (i == 0) {
+					points.write[index++] = circle_point - capsule_offset;
+				}
+				points.write[index++] = circle_point + capsule_offset;
+				if (i == first_half) {
+					points.write[index++] = circle_point - capsule_offset;
+					capsule_offset *= -1;
+				}
+			}
+		}
+	}
+	return points;
+}
+
+PackedVector2Array VisualShape2D::get_points() const {
+	return Transform2D(0, size / 2, 0, offset).xform(_get_shape_points());
+}
+
+PackedVector2Array VisualShape2D::get_uvs() const {
+	return Transform2D(0, Size2(0.5, 0.5), 0, Point2(0.5, 0.5)).xform(_get_shape_points());
+}
+
+Rect2 VisualShape2D::get_rect() const {
+	return Rect2(offset - size / 2, size);
+}
+
+void VisualShape2D::set_shape_type(ShapeType p_shape_type) {
+	if (shape_type == p_shape_type) {
+		return;
+	}
+	if (shape_type == SHAPE_CAPSULE) {
+		size.x *= 2;
+	}
+
+	shape_type = p_shape_type;
+	if (shape_type == SHAPE_CAPSULE) {
+		// Capsule default size should not have the same width and height.
+		size.x /= 2;
+	}
+	queue_redraw();
+	notify_property_list_changed();
+}
+
+VisualShape2D::ShapeType VisualShape2D::get_shape_type() const {
+	return shape_type;
+}
+
+void VisualShape2D::set_color(const Color &p_color) {
+	color = p_color;
```

Comment:
Missing check for the same color.

Reply:
Should I do the same for the other properties as well? `set_size`, `set_outline_width`, `set_antialiased`, etc.

Reply:
Yes.
Any setter that has side-effect (e.g. calls `queue_redraw()`) should be guarded.